presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java

```diff
@@ -0,0 +1,1994 @@
+        SearchRowIterator it = (SearchRowIterator) x.getLeft();
+
+        WrappedRowIterator x1;
+
+        Iterator<Row> h;
+        // optimize for count(*)
+        if (xp.isOnlyFetchRowCount()) {
+            h = new RowSearchIterator4CountStar(it);
+        }
+        else {
+            h = it;
+        }
+        x1 = new WrappedRowIterator(h, x.getMiddle(), x.getRight());
+
+        int firstFetchRows = 0;
+        Field field = ReflectionUtils.findField(SearchRowIterator.class, "result");
+        if (field != null) {
+            ReflectionUtils.makeAccessible(field);
+            SearchResponse grr = (SearchResponse) ReflectionUtils.getField(field, it);
+            if (grr != null && grr.getRows() != null) {
+                firstFetchRows = grr.getRows().size();
+            }
+        }
+        log.info(
+                "rowIterator4SearchIndex() end, queryId=%s firstFetchRows=%s totalMatchedCount=%s clientCost=%sms apiCost=%sms",
+                queryId,
+                firstFetchRows,
+                it.getTotalCount(),
+                x.getMiddle() / NANO_SECS_IN_ONE_MS,
+                x.getRight() / NANO_SECS_IN_ONE_MS);
+        return x1;
+    }
+
+    private WrappedRowIterator parallelScanResult(
+            String queryId,
+            TablestoreSplit split,
+            TablestoreColumnHandle[] columnHandles)
+    {
+        log.info("parallelScanResult start, queryId %s", queryId);
+        Map<String, Set<String>> candidateIndex = split.getIndexCandidates();
+        if (candidateIndex.isEmpty()) {
+            throw new IllegalArgumentException("Could not found suitable index in ParallelScan mode for query " + queryId);
+        }
+
+        // Columns to get
+        List<String> columns = Arrays.stream(columnHandles)
+                .map(TablestoreColumnHandle::getColumnName)
+                .collect(toList());
+
+        String indexName = null;
+        for (Map.Entry<String, Set<String>> entry : candidateIndex.entrySet()) {
+            if (entry.getValue().containsAll(columns)) {
+                indexName = entry.getKey();
+                break;
+            }
+        }
+        if (indexName == null) {
+            throw new IllegalArgumentException("Failed to find suitable index for query " + queryId + ". Please make sure all columns [" + columns +
+                    "] are in a single index, current indexes: " + candidateIndex);
+        }
+
+        TablestoreTableHandle tableHandle = split.getTableHandle();
+        TupleDomain<ColumnHandle> tupleDomain = split.getMergedTupleDomain();
+        String tableName = tableHandle.getTableName();
+
+        ParallelScanRequest parallelScanRequest = ParallelScanRequest.newBuilder()
+                .tableName(tableName)
+                .indexName(indexName)
+                .scanQuery(ScanQuery.newBuilder()
+                        .query(buildQueryForCurrentConditions(tupleDomain))
+                        .limit(2000)
```

Comment:
Just curious why the value here is 2000. Is it from the empirical experience?

Reply:
It is recommended by the tablestore team, it is not limiting the result count to 2000, it is limiting to only return 2000 for every request.